File: api/typescript/src/core/lib/enclaves/starlark_run_blocking.ts
import {Readable} from "stream";
import {
    StarlarkExecutionError,
    StarlarkInstruction,
    StarlarkInterpretationError, StarlarkRunResponseLine, StarlarkValidationError
} from "../../kurtosis_core_rpc_api_bindings/api_container_service_pb";

const STARLARK_RUN_OUTPUT_LINE_SPLIT = "\n"

export class StarlarkRunResult {
    constructor(
        public readonly runOutput: string,
        public readonly instructions: Array<StarlarkInstruction>,
        public readonly interpretationError: StarlarkInterpretationError | undefined,
        public readonly validationErrors: Array<StarlarkValidationError>,
        public readonly executionError: StarlarkExecutionError | undefined
    ){}
}

export async function readStreamContentUntilClosed(responseLines: Readable): Promise<StarlarkRunResult> {
    let scriptOutput = ""
    let interpretationError: StarlarkInterpretationError | undefined
    let validationErrors: Array<StarlarkValidationError> = []
    let executionError: StarlarkExecutionError | undefined
    let instructions: Array<StarlarkInstruction> = []

    return new Promise((resolve, error) => {
        responseLines.on('data', (responseLine: StarlarkRunResponseLine) => {
            if (responseLine.getInstruction() !== undefined) {
                instructions.push(responseLine.getInstruction()!)
            } else if (responseLine.getInstructionResult() !== undefined) {
                scriptOutput += responseLine.getInstructionResult()?.getSerializedInstructionResult() + STARLARK_RUN_OUTPUT_LINE_SPLIT
            } else if (responseLine.getError() !== undefined) {
                if (responseLine.getError()?.getInterpretationError() !== undefined) {
                    interpretationError = responseLine.getError()?.getInterpretationError()
                } else if (responseLine.getError()?.getValidationError() !== undefined) {
                    validationErrors.push(responseLine.getError()!.getValidationError()!)
                } else if (responseLine.getError()?.getExecutionError() !== undefined) {
                    executionError = responseLine.getError()?.getExecutionError()
                }
            } else if (responseLine.getRunFinishedEvent() !== undefined) {
                let runFinishedEvent = responseLine.getRunFinishedEvent()!
                if (runFinishedEvent.getIsrunsuccessful() && runFinishedEvent.getSerializedOutput() != "")  {
                        scriptOutput += runFinishedEvent.getSerializedOutput() + STARLARK_RUN_OUTPUT_LINE_SPLIT
                }
            }
        })
        responseLines.on('error', function () {
            if (!responseLines.destroyed) {
                responseLines.destroy();
                error(new Error("Unexpected error"))
            }
        });
        responseLines.on('end', function () {
            if (!responseLines.destroyed) {
                responseLines.destroy();
                resolve(new StarlarkRunResult(scriptOutput, instructions, interpretationError, validationErrors, executionError))
            }
        });
    })
}

File: api/golang/core/lib/enclaves/starlark_run_blocking.go
package enclaves

import (
	"github.com/kurtosis-tech/kurtosis/api/golang/core/kurtosis_core_rpc_api_bindings"
	"strings"
)

const (
	starlarkRunOutputLinesSplit = "\n"
)

type StarlarkRunMultilineOutput string

type StarlarkRunResult struct {
	RunOutput StarlarkRunMultilineOutput

	Instructions []*kurtosis_core_rpc_api_bindings.StarlarkInstruction

	InterpretationError *kurtosis_core_rpc_api_bindings.StarlarkInterpretationError

	ValidationErrors []*kurtosis_core_rpc_api_bindings.StarlarkValidationError

	ExecutionError *kurtosis_core_rpc_api_bindings.StarlarkExecutionError
}

func NewStarlarkRunResult(runOutput StarlarkRunMultilineOutput, instructions []*kurtosis_core_rpc_api_bindings.StarlarkInstruction, interpretationError *kurtosis_core_rpc_api_bindings.StarlarkInterpretationError, validationErrors []*kurtosis_core_rpc_api_bindings.StarlarkValidationError, executionError *kurtosis_core_rpc_api_bindings.StarlarkExecutionError) *StarlarkRunResult {
	return &StarlarkRunResult{
		RunOutput:           runOutput,
		Instructions:        instructions,
		InterpretationError: interpretationError,
		ValidationErrors:    validationErrors,
		ExecutionError:      executionError,
	}
}

func ReadStarlarkRunResponseLineBlocking(starlarkRunResponseLines <-chan *kurtosis_core_rpc_api_bindings.StarlarkRunResponseLine) *StarlarkRunResult {
	scriptOutput := strings.Builder{}
	instructions := make([]*kurtosis_core_rpc_api_bindings.StarlarkInstruction, 0)
	var interpretationError *kurtosis_core_rpc_api_bindings.StarlarkInterpretationError
	validationErrors := make([]*kurtosis_core_rpc_api_bindings.StarlarkValidationError, 0)
	var executionError *kurtosis_core_rpc_api_bindings.StarlarkExecutionError

	for responseLine := range starlarkRunResponseLines {
		if responseLine.GetInstruction() != nil {
			instructions = append(instructions, responseLine.GetInstruction())
		} else if responseLine.GetInstructionResult() != nil {
			scriptOutput.WriteString(responseLine.GetInstructionResult().GetSerializedInstructionResult())
			scriptOutput.WriteString(starlarkRunOutputLinesSplit)
		} else if responseLine.GetError() != nil {
			if responseLine.GetError().GetInterpretationError() != nil {
				interpretationError = responseLine.GetError().GetInterpretationError()
			} else if responseLine.GetError().GetValidationError() != nil {
				validationErrors = append(validationErrors, responseLine.GetError().GetValidationError())
			} else if responseLine.GetError().GetExecutionError() != nil {
				executionError = responseLine.GetError().GetExecutionError()
			}
		} else if responseLine.GetRunFinishedEvent() != nil {
			runFinishedEvent := responseLine.GetRunFinishedEvent()
			if runFinishedEvent.GetIsRunSuccessful() && runFinishedEvent.GetSerializedOutput() != "" {
				scriptOutput.WriteString(runFinishedEvent.GetSerializedOutput())
				scriptOutput.WriteString(starlarkRunOutputLinesSplit)
			}
		}
	}
	return NewStarlarkRunResult(
		StarlarkRunMultilineOutput(scriptOutput.String()),
		instructions,
		interpretationError,
		validationErrors,
		executionError)
}
